### PR TITLE
refactor class and name based resource name caching

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java [hunk 1]
```diff
@@ -7,18 +7,34 @@
...
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 
 public class AwsSdkClientDecorator extends HttpClientDecorator<Request, Response> {
 
   static final String COMPONENT_NAME = "java-aws-sdk";
 
-  private final Map<String, String> serviceNames = new ConcurrentHashMap<>();
-  private final Map<Class, String> operationNames = new ConcurrentHashMap<>();
+  private final QualifiedClassNameCache cache =
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java:21]: I noticed based on the contents of `serviceNames` and `operationNames` in a customer heap dump, and afterwards consulting the AWS SDK, that caching the resource name would be very low cardinality and a quick win.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java:21]: Nice catch!

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Function.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package datadog.trace.bootstrap.instrumentation.api;
+
+public interface Function<T, U> {
```
Comment (lpriima) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Function.java:3]: why you called this just `Function` ? 
In disregard your comment:
https://github.com/DataDog/dd-trace-java/pull/1759/files#diff-3438fdfbe14e3ba1c3a3e78ef37f7637R3

Comment (richardstartin) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/api/Function.java:3]: It's different. This is a function. We will search and replace `datadog.trace.bootstrap.instrumentation.api.Function` with `java.util.Function` when we drop JDK7. The other one differs to `BiFunction` in its interface.

**Pull request merged**